truth_set/python/structure/solution-approach/semantic-duplication/services/userService.py

```diff
@@ -0,0 +1,19 @@
+from datetime import datetime
+from ..utils.dateUtils import format_date
+
+class UserService:
+    def __init__(self):
+        self.users = []
+    
+    def add_user(self, name):
+        new_user = {
+            'id': len(self.users) + 1,
+            'name': name,
+            'registration_date': datetime.now()
+        }
+        self.users.append(new_user)
+        print(f"User Added: {new_user['name']}, Registered On: {format_date(new_user['registration_date'], 'MM/DD/YYYY')}")
```

Comment:
<div>


<div id="suggestion">
<div id="issue"><b>Consider abstracting common service functionality</b></div>
<div id="fix">

The `add_user` and `add_order` methods in `UserService` and `OrderService` appear to have similar functionality for managing records. Consider creating a base service class with common functionality to reduce semantic duplication. A similar issue was also found in truth_set/python/structure/solution-approach/semantic-duplication/index.py (line 1-17).
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<blockquote>Check the AI-generated fix before applying</blockquote>
<div id="code">


```suggestion
class BaseService:
    def __init__(self):
        self.records = []

    def add_record(self, data, entity_type):
        new_record = {
            'id': len(self.records) + 1,
            **data,
            'created_date': datetime.now()
        }
        self.records.append(new_record)
        return new_record

class UserService(BaseService):
    def add_user(self, name):
        data = {
            'name': name,
        }
        new_user = self.add_record(data, 'user')
        print(f"User Added: {new_user['name']}, Registered On: {format_date(new_user['registration_date'], 'MM/DD/YYYY')}")
```

</div>
</details>
</div>


<small><i>Code Review Run <a href=https://github.com/rkj6124/bito-test/pull/1#issuecomment-2536424992>#fdd196</a></i></small>
</div>

---
Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged